kratos/sources/model_part.cpp

```diff
@@ -13,21 +13,107 @@
 
 // System includes
 #include <sstream>
+#include <type_traits>
 
 // External includes
 
 // Project includes
 #include "includes/define.h"
 #include "includes/model_part.h"
 #include "includes/exception.h"
-#include "utilities/parallel_utilities.h"
+#include "utilities/reduction_utilities.h"
 
 namespace Kratos
 {
 
 KRATOS_CREATE_LOCAL_FLAG(ModelPart, ALL_ENTITIES, 0);
 KRATOS_CREATE_LOCAL_FLAG(ModelPart, OVERWRITE_ENTITIES, 1);
 
+namespace ModelPartHelperUtilities
+{
+
+template<class TContainerGetter>
+void AddEntitiesFromIds(
+    TContainerGetter&& rContainerGetter,
+    ModelPart* pModelPart,
+    const std::vector<IndexType>& rEntityIds)
+{
+    KRATOS_TRY
+
+    using container_type = std::remove_pointer_t<decltype(std::declval<TContainerGetter>()(std::declval<ModelPart*>()))>;
+
+    if(pModelPart->IsSubModelPart()) { //does nothing if we are on the top model part
+        //obtain from the root model part the corresponding list of nodes
+        ModelPart* root_model_part = &pModelPart->GetRootModelPart();
+        const auto& r_container = *rContainerGetter(root_model_part);
+
+        // we first sort and unique the given entity ids.
+        std::vector<typename container_type::pointer> entities;
+        entities.resize(rEntityIds.size());
+
+        // we are doing this in parallel because the most complex one is the finding of the entity for given id.
+        IndexPartition<IndexType>(rEntityIds.size()).for_each([pModelPart, &r_container, &entities,  &rEntityIds](const auto Index) {
+            const auto entity_id = rEntityIds[Index];
+
+            auto it = r_container.find(entity_id);
+
+            KRATOS_ERROR_IF(it == r_container.end())
+                << "while adding " << ModelPart::Container<container_type>::GetEntityName() << "s to submodelpart "
+                << pModelPart->FullName() << ", the "
+                << ModelPart::Container<container_type>::GetEntityName()
+                << " with Id " << entity_id << " does not exist in the root model part";
+
+            entities[Index] = *(it.base());
+        });
+
+        // aux is created to avoid doing sort and unique for all the parent model parts and current model part
+        // when insertion is done. since we dont use the entities anymore we can use the move operator here.
+        // since aux is an empty container. it will call push_back in the insert.
+        container_type aux;
+        aux.insert(std::move(entities));
```

Comment:
why do we need the "container_type" at all here? cannot we move directly the entities in line 78?